### PR TITLE
Allow using #[pin_project]  type with private field types

### PR DESCRIPTION
**Pull request by Aaron1011**
Previously, given code such as:

```rust

struct Private<T>;

#[pin_project]
pub struct Public<T> {
    #[pin] private: Private<T>
}
```

we would generate an Unpin impl like this:

```rust
impl Unpin for Public where Private: Unpin {}
```

Unfortunately, since Private is not a public type,
this would cause an E0446 ('private type `Private` in public interface)

When RFC 2145 is implemented (https://github.com/rust-lang/rust/issues/48054),
this will become a lint, rather then a hard error.

In the time being, we need a solution that will work with the current
type privacy rules.

The solution is to generate code like this:

```rust

fn __private_scope() {
    pub struct __UnpinPublic<T> {
        __field0: Private<T>
    }
    impl<T> Unpin for Public<T> where __UnpinPublic<T>: Unpin {}
}
```

That is, we generate a new struct, containing all of the pinned
fields from our #[pin_project] type. This struct is delcared within
a function, which makes it impossible to be named by user code.
This guarnatees that it will use the default auto-trait impl for Unpin -
that is, it will implement Unpin iff all of its fields implement Unpin.
This type can be safely declared as 'public', satisfiying the privacy
checker without actually allowing user code to access it.

This allows users to apply the #[pin_project] attribute to types
regardless of the privacy of the types of their fields.

### CODE REVIEW

File: pin-project-internal/src/pin_project/mod.rs [hunk 1]
```diff
@@ -109,21 +112,107 @@ impl Context {
...
+                    struct AlwaysUnpin<T: ?Sized> {
+                        val: ::core::marker::PhantomData<T>
+                    }
+
+                    impl<T: ?Sized> ::core::marker::Unpin for AlwaysUnpin<T> {}
+
+                    pub struct #struct_ident #full_generics #where_clause {
+                        __pin_project_use_generics: AlwaysUnpin<(#(#type_params),*)>,
+
+                        #(#fields),* #fields_comma
```
Comment (taiki-e) [inline pin-project-internal/src/pin_project/mod.rs]:
```suggestion
                        #(#fields,)*
```

File: tests/pin_project.rs [hunk 2]
```diff
@@ -238,3 +238,27 @@ fn combine() {
...
 }
+
+#[test]
+// This 'allow' is unrelated to the code
+// generated by pin-project - it's just to
+// allow us to put a private enum in a public enum
+#[allow(private_in_public)]
+fn test_private_type_in_public_type() {
+    #[pin_project]
+    pub struct PublicStruct<T> {
```
Comment (taiki-e) [inline tests/pin_project.rs]: My concern is that this generates the following documents (due to https://github.com/rust-lang/rust/issues/63281):
```rust
impl<T> Unpin for PublicStruct<T>
where
    __UnpinStructPublicStruct<T>: Unpin, 
```

Comment (taiki-e) [reply tests/pin_project.rs]: Even if `UnsafeUnpin` is used, a similar document is generated, but `UnsafeUnpin impl` is also documented at the same time, so we can know which type needs to be `Unpin`.
```rust
impl<T> Unpin for Foo<T>
where
    Wrapper<Self>: UnsafeUnpin,
```

**Comment by taiki-e** [comment]
@Aaron1011: Okay, thanks for the investigation.

> Since `docs.rs` uses nightly rust, libraries will automatically get the nicer documentation when built on `docs.rs`. Users building documentation locally can use `cargo +nightly doc`, but still target stable rust.

Should we provide a way to opt-out of this? (the documents generated this way is probably not very clean) cc @Nemo157

File: pin-project-internal/src/pin_project/mod.rs [hunk 3]
```diff
@@ -109,21 +112,133 @@ impl Context {
...
+
+            let inner_data = quote! {
+
+                struct #always_unpin_ident <T: ?Sized> {
+                    val: ::core::marker::PhantomData<T>
+                }
+
+                impl<T: ?Sized> ::core::marker::Unpin for #always_unpin_ident <T> {}
+
+                pub struct #struct_ident #full_generics #where_clause {
```
Comment (taiki-e) [inline pin-project-internal/src/pin_project/mod.rs]: Could you use struct/enum's visibility instead of `pub`? (add a field for this to Context)

**Comment by taiki-e** [comment]
bors delegate+

File: pin-project-internal/src/pin_project/mod.rs (hunk 3 above)
Comment (taiki-e) [reply pin-project-internal/src/pin_project/mod.rs]: Okay, thanks for the explanation.

**Comment by Aaron1011** [comment]
> Should we provide a way to opt-out of this? (the documents generated this way is probably not very clean)

I don't really see a use case for this. While the generated documentation certainly could be cleaner, I think it's always useful to know when a pin-projected type implements `Unpin`.

**Comment by Aaron1011** [comment]
@taiki-e: I'm added a comment explaining why the generated struct must be public, and squashed my commits.

**Comment by Aaron1011** [comment]
@taiki-e: Once https://github.com/rust-lang/rust/issues/48054 is implemented and stabilized, it will be possible to remove almost all of this code, and just generate normal bounds in the 'where' clause.

**Comment by taiki-e** [comment]
@Aaron1011: That's great, I'll look forward to that!

**Pull request merged**